README.md
# smooth-shadows

> A library to generate smooth shadows with multiple layers

## Install

```sh
npm install smooth-shadows
```

or

```sh
yarn add smooth-shadows
```

## Usage

```ts
import { smoothShadows } from 'smooth-shadows'

const shadows = smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08])

console.log(shadows)
/*
[
  [ 0, '0.216px', '1.361px', 0, [ 0, 0, 0, 0.027 ] ],
  [ 0, '0.622px', '3.197px', 0, [ 0, 0, 0, 0.039 ] ],
  [ 0, '1.447px', '6.325px', 0, [ 0, 0, 0, 0.049 ] ],
  [ 0, '4px', '16px', 0, [ 0, 0, 0, 0.08 ] ]
]
*/
```

## API

```ts
smoothShadows(x, y, blur, spread, color, options)
```

### Parameters

- `x`: offset x
  - Type: `string | 0`
- `y`: offset y
  - Type: `string | 0`
- `blur`: blur radius
  - Type: `string | 0`
- `spread`: spread radius
  - Type: `string | 0`
- `color`: rgba
  - Type: `[number, number, number, number]`
- `options` (optional)
  - `layers` (optional): number of layers
    - Type: `number`
    - Default: `4`
  - `easings` (optional):
    - `offset` (optional): easing of `x` and `y`, specify [easing functions](https://easings.net/) or [cubic-bezier](https://cubic-bezier.com/)
      - Type: `string | [number, number, number, number]`
      - Default: [`[0.65, 0.1, 0.9, 0.4]`](https://cubic-bezier.com/#.65,.1,.9,.4)
    - `blur` (optional): easing of blur, specify [easing functions](https://easings.net/) or [cubic-bezier](https://cubic-bezier.com/)
      - Type: `string | [number, number, number, number]`
      - Default: [`[0.65, 0.2, 0.9, 0.4]`](https://cubic-bezier.com/#.65,.2,.9,.4)
    - `alpha` (optional): easing of alpha, specify [easing functions](https://easings.net/) or [cubic-bezier](https://cubic-bezier.com/)
      - Type: `string | [number, number, number, number]`
      - Default: [`[0.2, 0.6, 0.9, 0.4]`](https://cubic-bezier.com/#.2,.6,.9,.4)
  - `output` (optional): return values
    - Type: `'array' | 'object' | 'css'`
    - Default: `'array'`

### Return values

Depends on the value of the output option.

<h4 id="api-return-values-array"><code>array</code> (default)</h4>

```json
[
  [0, "0.216px", "1.361px", 0, [0, 0, 0, 0.027]],
  [0, "0.622px", "3.197px", 0, [0, 0, 0, 0.039]],
  [0, "1.447px", "6.325px", 0, [0, 0, 0, 0.049]],
  [0, "4px", "16px", 0, [0, 0, 0, 0.08]]
]
```

<h4 id="api-return-values-object"><code>object</code></h4>

```json
[
  {
    "x": 0,
    "y": "0.216px",
    "blur": "1.361px",
    "spread": 0,
    "color": { "red": 0, "green": 0, "blue": 0, "alpha": 0.027 }
  },
  {
    "x": 0,
    "y": "0.622px",
    "blur": "3.197px",
    "spread": 0,
    "color": { "red": 0, "green": 0, "blue": 0, "alpha": 0.039 }
  },
  {
    "x": 0,
    "y": "1.447px",
    "blur": "6.325px",
    "spread": 0,
    "color": { "red": 0, "green": 0, "blue": 0, "alpha": 0.049 }
  },
  {
    "x": 0,
    "y": "4px",
    "blur": "16px",
    "spread": 0,
    "color": { "red": 0, "green": 0, "blue": 0, "alpha": 0.08 }
  }
]
```

<h4 id="api-return-values-css"><code>css</code></h4>

```txt
0 0.216px 1.361px rgba(0, 0, 0, 0.027), 0 0.622px 3.197px rgba(0, 0, 0, 0.039), 0 1.447px 6.325px rgba(0, 0, 0, 0.049), 0 4px 16px rgba(0, 0, 0, 0.08)
```

## Examples

### Increase layers

```ts
const shadows = smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { layers: 8 })

console.log(shadows.length) //=> 8
```

### Decrease layers

```ts
const shadows = smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { layers: 2 })

console.log(shadows.length) //=> 2
```

### Specify easing with easing functions

Can be specify the easing function of [easings.net](https://easings.net/). (But, cannot specify `ease*Elastic` and `ease*Bounce`)

```ts
smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], {
  easings: { offset: 'easeInExpo', blur: 'easeInQuart', alpha: 'easeInOutCirc' },
})
```

</details>

### Specify easing with cubic-bezier

Can be specify cubic-bezier, which was created by [cubic-bezier.com](https://cubic-bezier.com/).

```ts
smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], {
  easings: { offset: [0.65, 0.1, 0.9, 0.4], blur: [0.65, 0.2, 0.9, 0.4], alpha: [0.2, 0.6, 0.9, 0.4] },
})
```

### Output array

```ts
smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { output: 'array' })
```

[Return values](#api-return-values-array)

### Output object

```ts
smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { output: 'object' })
```

[Return values](#api-return-values-object)

### Output CSS

```ts
smoothShadows(0, '16px', '24px', 0, [0, 0, 0, 0.08], { output: 'css' })
```

[Return values](#api-return-values-css)

## Author

[@p-chan](https://github.com/p-chan)

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.1.0 (2021-08-09)


### Features

* improve arguments ([1a89ae6](https://github.com/p-chan/smooth-shadows/commit/1a89ae60bd68dcc6e8be26b8951a5ed9834d471b))
* improve types ([ce52434](https://github.com/p-chan/smooth-shadows/commit/ce52434a9e1319dde52cf00b7887206d325945fc))
* initial commit ([da7398f](https://github.com/p-chan/smooth-shadows/commit/da7398f575aa8e390044ec0289017dc80e0ef081))
* update default easings ([db173eb](https://github.com/p-chan/smooth-shadows/commit/db173eb624d5d88f9993042d8e5b9a0d5222a46d))
* update default output ([0c7b7fe](https://github.com/p-chan/smooth-shadows/commit/0c7b7fe1668780a6dd354a2447f74fea36678513))


### Bug Fixes

* fix function name ([653c07e](https://github.com/p-chan/smooth-shadows/commit/653c07ecb74f8a7ed9110fe2d02ef39b2188f48d))


test/index.test.ts
import { smoothShadows } from '../src'
import { testCasesGroups } from './test-cases-groups'

testCasesGroups.map(({ name, tests }) => {
  describe(name, () => {
    tests.map(({ name, input, output }) => {
      test(name, () => {
        const easingsOption = (() => {
          if (input.options == undefined || input.options.easings == undefined) return {}

          return input.options.easings
        })()

        const options = Object.assign({ output: 'css' as const }, { easings: easingsOption })

        expect(
          smoothShadows(
            input.x,
            input.y,
            input.blur,
            input.spread,
            [input.color[0], input.color[1], input.color[2], input.color[3]],
            options
          )
        ).toEqual(output)
      })
    })
  })
})

describe('layers options', () => {
  test('no option', () => {
    expect(smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08]).length == 4).toBeTruthy()
  })

  test('increase', () => {
    expect(smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { layers: 8 }).length == 8).toBeTruthy()
  })

  test('decrease', () => {
    expect(smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { layers: 2 }).length == 2).toBeTruthy()
  })
})

describe('output option', () => {
  test('no option', () => {
    expect(Array.isArray(smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08])[0])).toBeTruthy()
  })

  test('array', () => {
    expect(Array.isArray(smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { output: 'array' })[0])).toBeTruthy()
  })

  test('object', () => {
    expect(
      typeof smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { output: 'object' })[0] === 'object'
    ).toBeTruthy()
  })

  test('css', () => {
    expect(typeof smoothShadows(0, '4px', '16px', 0, [0, 0, 0, 0.08], { output: 'css' }) === 'string').toBeTruthy()
  })
})


src/shadows.ts
import BezierEasing from 'bezier-easing'
import parseUnit from 'parse-unit'
import roundTo from 'round-to'

import { CubicBezier, Easing, easingToCubicBezier } from './easings'

export type CSSLength = string | 0
export type CSSColorRGBA = [number, number, number, number]

export type SmoothShadowsArray = [CSSLength, CSSLength, CSSLength, CSSLength, CSSColorRGBA]
export type SmoothShadowsObject = {
  x: CSSLength
  y: CSSLength
  blur: CSSLength
  spread: CSSLength
  color: {
    red: number
    green: number
    blue: number
    alpha: number
  }
}
export type SmoothShadowsCSS = string

export type LayersOption = number
export type EasingsOption = Partial<{
  offset: Easing['name'] | CubicBezier
  blur: Easing['name'] | CubicBezier
  alpha: Easing['name'] | CubicBezier
}>
export type OutputOption = 'array' | 'object' | 'css'

function smoothShadows(
  x: CSSLength,
  y: CSSLength,
  blur: CSSLength,
  spread: CSSLength,
  color: CSSColorRGBA,
  options?: Partial<{ layers: LayersOption; easings: EasingsOption; output: 'array' }>
): SmoothShadowsArray[]

function smoothShadows(
  x: CSSLength,
  y: CSSLength,
  blur: CSSLength,
  spread: CSSLength,
  color: CSSColorRGBA,
  options?: Partial<{ layers: LayersOption; easings: EasingsOption; output: 'object' }>
): SmoothShadowsObject[]

function smoothShadows(
  x: CSSLength,
  y: CSSLength,
  blur: CSSLength,
  spread: CSSLength,
  color: CSSColorRGBA,
  options?: Partial<{ layers: LayersOption; easings: EasingsOption; output: 'css' }>
): SmoothShadowsCSS

function smoothShadows(
  x: CSSLength,
  y: CSSLength,
  blur: CSSLength,
  spread: CSSLength,
  color: CSSColorRGBA,
  options?: Partial<{ layers: LayersOption; easings: EasingsOption; output: OutputOption }>
): SmoothShadowsArray[] | SmoothShadowsObject[] | SmoothShadowsCSS {
  const [xValue, xUnit] = typeof x === 'string' ? parseUnit(x) : [x]
  const [yValue, yUnit] = typeof y === 'string' ? parseUnit(y) : [y]
  const [blurValue, blurUnit] = typeof blur === 'string' ? parseUnit(blur) : [blur]
  const [spreadValue, spreadUnit] = typeof spread === 'string' ? parseUnit(spread) : [spread]
  const [redValue, greenValue, blueValue, alphaValue] = color

  const layers = (options && options.layers) || 4
  const easings = {
    offset: options?.easings?.offset || [0.65, 0.1, 0.9, 0.4],
    blur: options?.easings?.blur || [0.65, 0.2, 0.9, 0.4],
    alpha: options?.easings?.alpha || [0.2, 0.6, 0.9, 0.4],
  }
  const output = (options && options.output) || 'array'

  const offsetBezierEasing = BezierEasing(
    ...(typeof easings.offset === 'string' ? easingToCubicBezier(easings.offset) : easings.offset)
  )
  const blurBezierEasing = BezierEasing(
    ...(typeof easings.blur === 'string' ? easingToCubicBezier(easings.blur) : easings.blur)
  )
  const alphaBezierEasing = BezierEasing(
    ...(typeof easings.alpha === 'string' ? easingToCubicBezier(easings.alpha) : easings.alpha)
  )

  const shadows: SmoothShadowsObject[] = [...Array(layers)].map((_, index) => {
    const progress = (index + 1) / layers

    return {
      x: xValue !== 0 ? `${roundTo(offsetBezierEasing(progress) * xValue, 3)}${xUnit}` : (0 as const),
      y: yValue !== 0 ? `${roundTo(offsetBezierEasing(progress) * yValue, 3)}${yUnit}` : (0 as const),
      blur: blurValue !== 0 ? `${roundTo(blurBezierEasing(progress) * blurValue, 3)}${blurUnit}` : (0 as const),
      spread: spreadValue !== 0 ? `${roundTo(spreadValue, 3)}${spreadUnit}` : (0 as const),
      color: {
        red: redValue,
        green: greenValue,
        blue: blueValue,
        alpha: roundTo(alphaBezierEasing(progress) * alphaValue, 3),
      },
    }
  })

  if (output === 'css') {
    return shadows
      .map((shadow) => {
        return `${shadow.x} ${shadow.y} ${shadow.blur}${shadow.spread ? ` ${shadow.spread}` : ''} rgba(${
          shadow.color.red
        }, ${shadow.color.green}, ${shadow.color.blue}, ${shadow.color.alpha})`
      })
      .join(', ')
  }

  if (output === 'object') return shadows

  return shadows.map(({ x, y, blur, spread, color }) => {
    return [x, y, blur, spread, [color.red, color.green, color.blue, color.alpha]]
  })
}

export { smoothShadows }


scripts/create-test-cases-groups.ts
import { format, resolveConfig } from 'prettier'
import { resolve } from 'path'
import { writeFile } from 'fs/promises'

import { easings } from '../src/easings'
import { smoothShadows, CSSLength, CSSColorRGBA, LayersOption, EasingsOption, OutputOption } from '../src/shadows'

export type TestCasesGroup = {
  name: string
  tests: TestCase[]
}

type TestCase = {
  name: string
  input: {
    x: CSSLength
    y: CSSLength
    blur: CSSLength
    spread: CSSLength
    color: CSSColorRGBA
    options?: Partial<{ layers: LayersOption; easings: EasingsOption; output: OutputOption }>
  }
  output: string
}

type TestInputCasesGroup = {
  name: string
  tests: TestInputCase[]
}

type TestInputCase = Omit<TestCase, 'output'>
;(async () => {
  const easingOptionValues: ['offset', 'blur', 'alpha'] = ['offset', 'blur', 'alpha']

  const testInputGroups: TestInputCasesGroup[] = [
    {
      name: 'basic',
      tests: [
        {
          name: 'basic',
          input: {
            x: 0,
            y: '4px',
            blur: '16px',
            spread: 0,
            color: [0, 0, 0, 0.08],
            options: { output: 'css' },
          },
        },
        {
          name: 'basic-with-spread',
          input: {
            x: 0,
            y: '4px',
            blur: '16px',
            spread: '64px',
            color: [0, 0, 0, 0.08],
            options: { output: 'css' },
          },
        },
      ],
    },
    {
      name: 'colors',
      tests: [
        // red
        {
          name: 'red',
          color: [255, 0, 0],
        },
        // green
        {
          name: 'green',
          color: [0, 255, 0],
        },
        // blue
        {
          name: 'blue',
          color: [0, 0, 255],
        },
        // white
        {
          name: 'white',
          color: [255, 255, 255],
        },
      ].map(({ name, color }) => {
        return {
          name: `color-${name}`,
          input: {
            x: 0,
            y: '4px',
            blur: '16px',
            spread: 0,
            color: [color[0], color[1], color[2], 0.08],
            options: { output: 'css' },
          },
        }
      }),
    },
    {
      name: 'easings',
      tests: easingOptionValues
        .map((option) => {
          const cubicBeziers: TestInputCase = {
            name: `easings-${option}-cubicBezierLinear`,
            input: {
              x: 0,
              y: '4px',
              blur: '16px',
              spread: 0,
              color: [0, 0, 0, 0.08],
              options: {
                easings: {
                  [option]: [0.25, 0.25, 0.75, 0.75],
                },
                output: 'css',
              },
            },
          }

          const easingFunction: TestInputCase[] = easings.map(({ name }) => {
            return {
              name: `esings-${option}-${name}`,
              input: {
                x: 0,
                y: '4px',
                blur: '16px',
                spread: 0,
                color: [0, 0, 0, 0.08],
                options: {
                  easings: {
                    [option]: name,
                  },
                  output: 'css',
                },
              },
            }
          })

          return [cubicBeziers, ...easingFunction]
        })
        .flat(),
    },
  ]

  const testCasesGroups: TestCasesGroup[] = testInputGroups.map(({ name, tests }) => {
    return {
      name,
      tests: tests.map(({ name, input }) => {
        const easingsOption = (() => {
          if (input.options == undefined || input.options.easings == undefined) return {}

          return input.options.easings
        })()

        const options = Object.assign({ output: 'css' as const }, { easings: easingsOption })

        return {
          name,
          input,
          output: smoothShadows(input.x, input.y, input.blur, input.spread, input.color, options),
        }
      }),
    }
  })

  const path = resolve(process.cwd(), 'test', 'test-cases-groups.ts')

  resolveConfig(resolve(process.cwd(), '.prettierrc')).then(async (options) => {
    if (options == undefined) throw new Error('options are not defined')

    await writeFile(
      path,
      format(
        `// Auto generate by ../scripts/create-test-cases-groups.ts\nimport { TestCasesGroup } from '../scripts/create-test-cases-groups'\n\nexport const testCasesGroups: TestCasesGroup[] = ${JSON.stringify(
          testCasesGroups
        )}`,
        options
      ),
      'utf-8'
    )
  })
})()


src/easings.ts
export type CubicBezier = [number, number, number, number]

export type Easing = {
  name:
    | 'easeInSine'
    | 'easeOutSine'
    | 'easeInOutSine'
    | 'easeInCubic'
    | 'easeOutCubic'
    | 'easeInOutCubic'
    | 'easeInQuint'
    | 'easeOutQuint'
    | 'easeInOutQuint'
    | 'easeInCirc'
    | 'easeOutCirc'
    | 'easeInOutCirc'
    | 'easeInQuad'
    | 'easeOutQuad'
    | 'easeInOutQuad'
    | 'easeInQuart'
    | 'easeOutQuart'
    | 'easeInOutQuart'
    | 'easeInExpo'
    | 'easeOutExpo'
    | 'easeInOutExpo'
    | 'easeInBack'
    | 'easeOutBack'
    | 'easeInOutBack'
  cubicBezier: CubicBezier
}

/**
 * @see https://easings.net/
 */
export const easings: Easing[] = [
  {
    name: 'easeInSine',
    cubicBezier: [0.12, 0, 0.39, 0],
  },
  {
    name: 'easeOutSine',
    cubicBezier: [0.61, 1, 0.88, 1],
  },
  {
    name: 'easeInOutSine',
    cubicBezier: [0.37, 0, 0.63, 1],
  },
  {
    name: 'easeInCubic',
    cubicBezier: [0.32, 0, 0.67, 0],
  },
  {
    name: 'easeOutCubic',
    cubicBezier: [0.33, 1, 0.68, 1],
  },
  {
    name: 'easeInOutCubic',
    cubicBezier: [0.65, 0, 0.35, 1],
  },
  {
    name: 'easeInQuint',
    cubicBezier: [0.64, 0, 0.78, 0],
  },
  {
    name: 'easeOutQuint',
    cubicBezier: [0.22, 1, 0.36, 1],
  },
  {
    name: 'easeInOutQuint',
    cubicBezier: [0.83, 0, 0.17, 1],
  },
  {
    name: 'easeInCirc',
    cubicBezier: [0.55, 0, 1, 0.45],
  },
  {
    name: 'easeOutCirc',
    cubicBezier: [0, 0.55, 0.45, 1],
  },
  {
    name: 'easeInOutCirc',
    cubicBezier: [0.85, 0, 0.15, 1],
  },
  {
    name: 'easeInQuad',
    cubicBezier: [0.11, 0, 0.5, 0],
  },
  {
    name: 'easeOutQuad',
    cubicBezier: [0.5, 1, 0.89, 1],
  },
  {
    name: 'easeInOutQuad',
    cubicBezier: [0.45, 0, 0.55, 1],
  },
  {
    name: 'easeInQuart',
    cubicBezier: [0.5, 0, 0.75, 0],
  },
  {
    name: 'easeOutQuart',
    cubicBezier: [0.25, 1, 0.5, 1],
  },
  {
    name: 'easeInOutQuart',
    cubicBezier: [0.76, 0, 0.24, 1],
  },
  {
    name: 'easeInExpo',
    cubicBezier: [0.7, 0, 0.84, 0],
  },
  {
    name: 'easeOutExpo',
    cubicBezier: [0.16, 1, 0.3, 1],
  },
  {
    name: 'easeInOutExpo',
    cubicBezier: [0.87, 0, 0.13, 1],
  },
  {
    name: 'easeInBack',
    cubicBezier: [0.36, 0, 0.66, -0.56],
  },
  {
    name: 'easeOutBack',
    cubicBezier: [0.34, 1.56, 0.64, 1],
  },
  {
    name: 'easeInOutBack',
    cubicBezier: [0.68, -0.6, 0.32, 1.6],
  },
]

export const easingToCubicBezier = (name: Easing['name']): CubicBezier => {
  const easing = easings.find((easing) => {
    return easing.name === name
  })

  if (easing == undefined) return [0, 0, 1, 1]

  return easing.cubicBezier
}
